rhino/src/main/java/org/mozilla/javascript/NativeSet.java

```diff
@@ -232,4 +296,473 @@ private static NativeSet realThis(Scriptable thisObj, String name) {
+        }
+        if (!(keysVal instanceof Callable)) {
+            throw ScriptRuntime.typeErrorById(
+                    "msg.isnt.function", "keys", ScriptRuntime.typeof(keysVal));
+        }
+
+        Callable hasMethod = (Callable) hasVal;
+        Callable keysMethod = (Callable) keysVal;
+
+        // ES2025: Compare sizes to determine iteration strategy
+        double otherSizeDouble = ScriptRuntime.toNumber(sizeVal);
+        if (Double.isNaN(otherSizeDouble)) {
+            throw ScriptRuntime.typeError("size is not a number");
+        }
+        int otherSize =
+                Double.isInfinite(otherSizeDouble)
+                        ? Integer.MAX_VALUE
+                        : (int) Math.floor(otherSizeDouble);
+        int thisSize = entries.size();
+
+        if (thisSize <= otherSize) {
+            // When this.size <= other.size: iterate through this, call other.has()
+            for (Hashtable.Entry entry : entries) {
+                Object key = entry.key;
+                Object inOther = callHas(cx, scope, otherObj, hasMethod, key);
+                if (ScriptRuntime.toBoolean(inOther)) {
+                    result.js_add(key);
+                }
+            }
+        } else {
+            // When this.size > other.size: iterate through other.keys(), call this.has()
+            Object iterator =
+                    ScriptRuntime.callIterator(
+                            keysMethod.call(
+                                    cx,
+                                    scope,
+                                    ScriptableObject.ensureScriptable(otherObj),
+                                    ScriptRuntime.emptyArgs),
+                            cx,
+                            scope);
+            try (IteratorLikeIterable it = new IteratorLikeIterable(cx, scope, iterator)) {
+                for (Object key : it) {
+                    if (js_has(key) == Boolean.TRUE) {
+                        result.js_add(key);
+                    }
+                }
+            }
+        }
+
+        return result;
+    }
+
+    private Object js_union(Context cx, Scriptable scope, Object[] args) {
+        Object otherObj = args.length > 0 ? args[0] : Undefined.instance;
+
+        NativeSet result = (NativeSet) cx.newObject(scope, CLASS_NAME);
+        result.instanceOfSet = true;
+
+        // Add all elements from this set
+        for (Hashtable.Entry entry : entries) {
+            result.js_add(entry.key);
+        }
+
+        // ES2025: GetSetRecord requires size, has, and keys properties
+        Scriptable scriptable = ScriptableObject.ensureScriptable(otherObj);
+        Object sizeVal = ScriptableObject.getProperty(scriptable, "size");
+        Object hasVal = ScriptableObject.getProperty(scriptable, "has");
+        Object keysVal = ScriptableObject.getProperty(scriptable, "keys");
+
+        // Validate all required properties exist
+        validateSetLike(sizeVal, hasVal, keysVal);
```

Comment:
why not doing

        Object sizeVal = ScriptableObject.getProperty(scriptable, "size");
        Object hasVal = ScriptableObject.getProperty(scriptable, "has");
        Object keysVal = ScriptableObject.getProperty(scriptable, "keys");

also inside the validateSetLike() method